src/ontodia/widgets/searchResults.tsx

```diff
@@ -0,0 +1,248 @@
+        const getNewSelection = (state: SearchResultsState): Dictionary<true> => {
+            let selection: Dictionary<true> = {...state.selection};
+            const previouslySelected = Boolean(selection[model.id]);
+            const items = this.props.items || [];
+            const modelIndex = items.indexOf(model);
+
+            if (event.shiftKey && this.focusIndex !== -1) { // select range
+                const startIndex = Math.min(this.focusIndex, modelIndex);
+                const finishIndex = Math.max(this.focusIndex, modelIndex);
+                selection = this.getSelectedRange(startIndex, finishIndex);
+            } else {
+                this.selectionIndex = this.focusIndex = modelIndex;
+                const ctrlKey = event.ctrlKey || event.metaKey;
+
+                if (ctrlKey) { // select/deselect
+                    event.preventDefault();
+                    if (previouslySelected && selection[model.id]) {
+                        delete selection[model.id];
+                    } else if (!previouslySelected && !selection[model.id]) {
+                        selection[model.id] = true;
+                    }
+                } else { // single click
+                    selection = {[model.id]: true};
+                }
+            }
+            return selection;
+        };
+
+        if (this.props.insteadOfUpdateSelection) {
+            this.props.insteadOfUpdateSelection(getNewSelection(this.state));
+        } else {
+            this.setState((state): SearchResultsState => {
+                return { selection: getNewSelection(state) };
+            });
+        }
+    }
+
+    private getSelectedRange(startIndex: number, finishIndex: number): Dictionary<true> {
+        const items = this.props.items || [];
+        const selection: Dictionary<true> = {};
+
+        for (let i = startIndex; i <= finishIndex; i++) {
+            const selectedModel = items[i];
+            if (!this.isOnDiagram(selectedModel)) {
+                selection[selectedModel.id] = true;
+            }
+        }
+        return selection;
+    }
+
+    private getNextIndex(curIndex: number, direction: 'up' | 'down') {
+        const items = this.props.items;
+        const step = direction === 'up' ? -1 : 1;
+        for (let i = curIndex + step; i !== curIndex; i += step) {
+            if (i < 0) {
+                i = items.length - 1;
+            }
+            if (i >= items.length) {
+                i = 0;
+            }
+            if (!this.isOnDiagram(items[i])) {
+                return i;
+            }
+        }
+        return 0;
+    }
+
+    private isOnDiagram(model: ElementModel): boolean {
+        return this.props.view.model.elements.findIndex(
+            element => element.iri === model.id && element.group === undefined
+        ) >= 0;
+    }
+
+    private focusOn(element: ElementModel) {
+        // const elementId = listElementViewId(element);
```

Comment:
I would like to discuss this place. So I kept some code here. ScrollIntoView works not so perfect when the component is builded in the connection menu.